optimism/src/mips/interpreter.rs

```diff
@@ -1060,7 +1061,157 @@ pub fn interpret_rtype<Env: InterpreterEnv>(env: &mut Env, instr: RTypeInstructi
             return;
         }
         RTypeInstruction::SyscallWriteHint => (),
-        RTypeInstruction::SyscallWritePreimage => (),
+        RTypeInstruction::SyscallWritePreimage => {
+            let addr = env.read_register(&Env::constant(5));
+            let write_length = env.read_register(&Env::constant(6));
+
+            // Cannon assumes that the remaining `byte_length` represents how much remains to be
+            // read (i.e. all write calls send the full data in one syscall, and attempt to retry
+            // with the rest until there is a success). This also simplifies the implementation
+            // here, so we will follow suit.
+            let bytes_to_preserve_in_register = {
+                // FIXME: Requires a range check
+                let pos = env.alloc_scratch();
+                unsafe { env.bitmask(&write_length, 2, 0, pos) }
+            };
+            let register_idx = {
+                let registers_left_to_write_after_this = {
+                    // FIXME: Requires a range check
+                    let pos = env.alloc_scratch();
+                    // The virtual register is 32 bits wide, so we can just read 6 bytes. If the
+                    // register has an incorrect value, it will be unprovable and we'll fault.
+                    unsafe { env.bitmask(&write_length, 6, 2, pos) }
```

Comment:
Trying to understand this.

Reply:
The write syscall will request 32 bytes, but we want to write aligned-word-wise.

Reply:
I wonder if what's being written is the claimed hash output, whereas it might be SyscallReadPreimage what reads any arbitrary length.